WIP: AMD Family 15h model 30h support

pipcet opened the pull request:
* support using the difference of two PMCs as tick count
* support AMD family 15h model 30h CPUs that way.

I revisited the issue of running rr on my AMD bdver3 (family 15h, model 30h) laptop, which I unfortunately appear  to be stuck with for a while. I'm not sure whether it's due to a microcode update, but it appears that there are now two performance counters whose difference is predictable enough to run rr:
0xc4 (taken branches including exceptions, interrupts, far branches, and syscalls) and 0xc6 (exceptions, interrupts, far branches, and syscalls).

The difference, of course, is the count of taken near branches, conditional or not. I think that's good enough for us.

Most tests pass without a syscallbuf, and those that do not appear to fail for unrelated reasons.  I'm still looking at the syscallbuf code to see how it breaks.

This is just a snapshot for now; I think the hard part was working out which counters to use. Any preferences for how best to submit this?

Both PMCs I'm using appear to be unchanged for family 17h, which I think is Ryzen, so it's possible this approach would also work for that CPU.

rocallahan commented:
> Any preferences for how best to submit this?

Not sure what you mean. The commit you've added looks pretty reasonable.

This is a really interesting idea. Is there a way to check whether this approach handles SMM interrupts? I recall they get counted as branches sometimes.

Can you confirm that the values you get by taking the difference actually correspond to the number of near calls retired?

pipcet commented:
> Is there a way to check whether this approach handles SMM interrupts? I recall they get counted as branches sometimes.

There's a separate PMC for SMIs, but I'm not seeing any hits there; however, if I'm reading the documentation correctly, both PMCs C4 and C6 would be incremented for an SMI, so the difference would remain constant.

> Can you confirm that the values you get by taking the difference actually correspond to the number of near calls retired?

Well, it's near taken branches (I think that's call, ret, jmp, and jCC if taken).

On assembly files it appears to work out the way it should; I'm not aware of a freely-available emulator that has a convenient way to count only taken branches, so I can't really test it works on large binaries. If you have any ideas for how to test this further, I'd appreciate the advice.

Right now I'm running a 2^32-jump loop in a loop to see whether there's any spurious deviations, but so far there aren't any.

I should point out that AMD's documentation makes no promises about the counters being exact, it's just experimental data.

rocallahan commented:
> I should point out that AMD's documentation makes no promises about the counters being exact, it's just experimental data.

That's normal.

zairil commented:
So sweet

pipcet commented:
Okay, I briefly had access to an Intel system, and programmed it to use counter 0xc4 with a umask of 0x30. That shows the same symptoms as my patch does on AMD: it counts all taken branches, which works fine without a syscallbuf, but breaks with one.

For example

```
int main(void)
{
  for (int i = 0; i < 128*128; i++)
    lseek(0, 0, 0);
}
```

will cause the problem.

pipcet commented:
Okay, I found the culprit. It's this code in `record_signal.cc`:
```
    // This is a single instruction that jumps to the location stored in
    // preload_thread_locals::stub_scratch_1. Emulate it.
    Registers r = t->regs();
    r.set_ip(get_stub_scratch_1(t));
    t->set_regs(r);
```
Which obviously needs to increment `t`'s ticks if we're counting all taken branches. Fixing that leaves 21 (of 2253) test failures, which I'm currently looking at.

rocallahan commented:
Good catch!

rocallahan commented:
One thing we'll have to deal with here, assuming you work out the test failures, is that in the `rr_trace.capnp` `Header` we'll have to add a new field indicating what "ticks" mean in this trace. In the initial landing I guess we'll refuse to replay a trace if the meaning of "ticks" is different in the host system than indicated in the trace. It sounds like as an extra feature we might be able to reconfigure (some?) Intel CPUs to use the same meaning as AMD ... and maybe some future AMD CPUs could be configured to support RCB.

zairil commented:
So Sweet

pipcet commented:
```
100% tests passed, 0 tests failed out of 2253
```
admittedly with at least one workaround for an mmap problem I don't understand (I think it's unrelated to the PMC code, possibly a kernel configuration issue). The code's at https://github.com/mozilla/rr/compare/master...pipcet:alltestspass?expand=1.

rocallahan commented:
Great! I left some comments on the commits. I only saw one workaround though; where's the other one?

rocallahan commented:
As a followup to this, it would definitely be worth exploring whether Intel CPUs can operate in the same mode. If they can, that would be better than using RCB because taken branches are more dense: we sometimes have problems with long code sequences of nested function calls and loops not involving conditional branches.

OTOH in some cases using two performance counters instead of one will be a disadvantage. E.g. in Mac VMWare guests we sometimes only have one usable counter.

rocallahan commented:
A couple more thoughts:

We need to be careful about reusing a ticks-semantics-ID across architectures to make sure that the semantics really are identical across those architectures. Multiple IDs that turn out to be the same semantics would be tolerable, but using the same ID where it turns out the semantics are subtly different would be a problem. (I suppose we could recover with CPUID checks but that would add complexity.)

It would be good to check to make sure an indirect branch with a memory operand that triggers a page fault or signal and gets restarted does not get counted twice. Ideally we'd have a test for this.

rocallahan commented:
Another thing that would be interesting to know is whether rr works in a KVM guest with perfcounter virtualization.

pipcet commented:
@rocallahan Thanks for the review and the suggestions!

I've rebased and updated the c4c6 branch, and now it's passing all tests except for the "mremap" ones, which run into an assert condition even if I run `./bin/mremap` directly without using rr. That appears to happen on Ubuntu and Fedora live images run in KVM, too, so I'm not sure whether it isn't simply a (recent?) change in Linux behavior...

pipcet commented:
I've just run the 64-bit tests on a KVM guest, and the only failures were mremap and prctl_speculation_ctrl (which asserts that prctl returns ENXIO when it actually returns EINVAL).

rocallahan commented:
> I've just run the 64-bit tests on a KVM guest, and the only failures were mremap and prctl_speculation_ctrl (which asserts that prctl returns ENXIO when it actually returns EINVAL).

Great!

rocallahan commented:
> now it's passing all tests except for the "mremap" ones, which run into an assert condition even if I run ./bin/mremap directly without using rr.

I'm not seeing this with kernel 4.17.19-200.fc28.x86_64. What's your kernel?

rocallahan commented:
> What's invalid about having mremap reuse and extend our existing mapping?

Theoretically, nothing. We should remove that assertion and the corresponding one in `mremap_non_page_size`. I'll do it.

In practice I've always observed Linux allocating fresh mmaps top-down, so there isn't enough room to expand the wpage mapping and it has to be allocated at a fresh address. I'm surprised that's changed, but OK.

rocallahan commented:
> I'll do it.

Done.

rocallahan commented:
I looked at https://github.com/mozilla/rr/commit/8f5a8b5ed63d2389da81865c6f3c0931abb332f0#r30518535 again.

> that means we can't coalesce the overflow of the initial mapping with the second-to-last mmap, even though we could coalesce them during recording.

So this is the `checksum_sanity.run` test, right?

The mmaps I get running that test locally are a bit simpler than your log showed, so this is probably an existing rr bug that happens to be triggered by your Debian build.

I will try modifying `checksum_sanity` to reproduce the bug.

rocallahan commented:
In the meantime if that's your only test failure then I think it's safe to assume it's unrelated to CPU architecture and we should proceed with the rest of your patches.

rocallahan commented on src/PerfCounters.cc:
```diff
@@ -556,7 +578,9 @@ static void init_attributes() {
   hw_interrupts_attr.exclude_hv = 1;
 
-  check_for_bugs();
+  if (ticks_for_branch == 0) {
```
Overloading `ticks_for_branch` this way isn't great.

I think this might be a good time to switch the two bools and `ticks_for_branch` in `PmuConfig` to a flags word with flags, say, `FLAG_UNSUPPORTED`, `FLAG_BENEFITS_FROM_USELESS_COUNTER`, `FLAG_CHECK_FOR_INTEL_BUGS`, and `FLAG_UNCONDITIONAL_INDIRECT_JUMP_ADDS_TICK`.

rocallahan replied:
Or since this is local to `PerfCounters.cc` we can probably skip the `FLAG_` prefix.

pipcet replied:
> Overloading `ticks_for_branch` this way isn't great.

You're right. It's appropriate for check_working_counter, I guess, but not for the others.

> I think this might be a good time to switch the two bools and `ticks_for_branch` in `PmuConfig` to a flags word with flags, say, `FLAG_UNSUPPORTED`, `FLAG_BENEFITS_FROM_USELESS_COUNTER`, `FLAG_CHECK_FOR_INTEL_BUGS`, and `FLAG_UNCONDITIONAL_INDIRECT_JUMP_ADDS_TICK`.

I don't like the idea of flag words much (as opposed to structs of booleans), but okay.

rocallahan replied:
The advantage of a flags word is that you get a default value of 0 for flags you don't explicitly set.

rocallahan commented on src/PerfCounters.cc:
```diff
@@ -575,7 +599,21 @@ bool PerfCounters::is_ticks_attr(const perf_event_attr& attr) {
+    return true;
+  }
+  return false;
+}
```
Why did you change this?

pipcet replied:
Thanks for catching this. I'd simply forgotten to revert to the old code after briefly returning true for ticks_attr and minus_ticks_attr.

rocallahan commented on src/VirtualPerfCounterMonitor.h:
```diff
@@ -18,6 +22,7 @@ namespace rr {
 class VirtualPerfCounterMonitor : public FileMonitor {
 public:
   static bool should_virtualize(const struct perf_event_attr& attr);
+  static VirtualPerfCounterType virtualization_type(const struct perf_event_attr& attr);
```
Did you see where I suggested instead of doing this, create a synthetic perf counter ID that we use on all architectures that always counts `ticks`, and use that when running rr replay under rr?

pipcet replied:
Yes, and I think it's a good idea, I just haven't gotten around to it yet. I'd suggest simply not virtualizing performance counters on AMD CPUs until we introduce that.

rocallahan replied:
If you don't virtualize at all then the rr under rr test (`record_replay.run`) will fail.

rocallahan commented on src/PerfCounters.h:
```diff
@@ -66,6 +66,12 @@ class PerfCounters {
+   * Return the number of ticks we need for an emulated branch.
+   * `t` is used for debugging purposes.
+   */
+  static Ticks ticks_for_branch(Task* t);
```
I think we need to specify here that this is an unconditional, indirect branch.

rocallahan commented:
To be clear, please remove the "fix and simplify the code for mmaps extending past EOF" patch from this set and we'll deal with that as a separate issue.

rocallahan commented:
I've filed https://github.com/mozilla/rr/issues/2261 for that issue.

pipcet commented:
> To be clear, please remove the "fix and simplify the code for mmaps extending past EOF" patch from this set and we'll deal with that as a separate issue.

Sure! Should I also remove the performance counter virtualization until we can fix that properly (one way or the other)?

rocallahan commented:
> Should I also remove the performance counter virtualization until we can fix that properly (one way or the other)?

I'd prefer to have that here since that's not a previously existing bug.

pipcet commented:
Okay, so this version has the virtualization (with 0xc6 always reading as 0) and AMD support.

rocallahan commented:
> Okay, so this version has the virtualization (with 0xc6 always reading as 0) and AMD support.

Github's review comment tracking is awful so maybe you missed this, but I'd prefer to define a synthetic performance counter value, the same value for all architectures, that always maps to "ticks", virtualize that under rr, and always use that when rr replay is running under rr.

I think that will be a little simpler, but more importantly, the same code will be used on both Intel and AMD, which reduces the likelihood of accidentally breaking one of them.

pipcet commented:
Sorry. No, I didn't miss it, I just don't know how it's done so I wanted to get the other changes out of the way first.

I'll look into this.

pipcet commented:
So, do you want to usurp a hardware performance counter ID for this, or implement it as a software performance event?

rocallahan commented:
Whichever is easier.

pipcet commented:
I've got something working; I made the VPMC have no skid, by single-stepping when we get close to the target tick value. That seemed better to me than exposing the actual PMC's skid_size to the VPMC user somehow.

https://github.com/pipcet/rr/compare/c4c6...pipcet:c4c6-virtperf?expand=1

rocallahan commented on src/VirtualPerfCounterMonitor.h:
```diff
@@ -47,6 +52,7 @@ class VirtualPerfCounterMonitor : public FileMonitor {
   void maybe_enable_interrupt(Task* t, uint64_t after);
   void disable_interrupt() const;
 
+  VirtualPerfCounterType pmctype_;
```
We don't need `VirtualPerfCounterType` anymore, right?

rocallahan commented:
> I'm seeing test failures in some of the mmap tests, but I think that's due to your recent changes.

Could be, but 100% of tests are passing for me.

pipcet commented:
Thanks! I've rebased with the changes you requested.

rocallahan commented:
Thanks!

rocallahan commented:
FYI I added a `TicksSemantics` field to the rr trace header to track and check what sort of ticks were used. https://github.com/mozilla/rr/commit/a796f774b1eb9c7a08479faaa90ff8b55ac70c36
